### PR TITLE
Refactored indexing to work on a per-field basis.

### LINKED ISSUE #18: Create an index per field
Right now we create a combined index across all text fields. In the case of metadata, we create an index of an index. We need to store an index per field for better search resolution.

### PR DESCRIPTION
**Pull request by krisrogers**
Closes #18

### CODE REVIEW

**Comment by rstuart85** [comment]
@krisrogers tests aren't passing.

**Comment by krisrogers** [comment]
@rstuart85 updated

File: caterpillar/processing/index.py [hunk 1]
```diff
@@ -302,15 +305,18 @@ def begin(self, timeout=None):
                 storage = self.__config.storage_cls(self._path, create=True)
                 # Need to create the containers
                 storage.begin()
+                for field_name in self.__schema.get_indexed_text_fields():
```
Comment (rstuart85) [inline caterpillar/processing/index.py]: This should just check `text_fields` you created on line 262 right? Just saves a function call.

File: caterpillar/processing/index.py [hunk 2]
```diff
@@ -302,15 +305,18 @@ def begin(self, timeout=None):
...
+                    storage.add_container(IndexWriter.FRAMES_CONTAINER.format(field_name))
+                    storage.add_container(IndexWriter.POSITIONS_CONTAINER.format(field_name))
+                    storage.add_container(IndexWriter.ASSOCIATIONS_CONTAINER.format(field_name))
+                    storage.add_container(IndexWriter.FREQUENCIES_CONTAINER.format(field_name))
+                    storage.add_container(IndexWriter.METADATA_CONTAINER.format(field_name))
                 storage.add_container(IndexWriter.DOCUMENTS_CONTAINER)
-                storage.add_container(IndexWriter.FRAMES_CONTAINER)
+                storage.add_container(IndexWriter.FRAMES_CONTAINER.format(None))  # non-text frames
```
Comment (rstuart85) [inline caterpillar/processing/index.py]: This is a cool solution. Nice work!

File: caterpillar/processing/index.py [hunk 3]
```diff
@@ -338,33 +344,37 @@ def flush(self):
         """
         logger.debug("Flushing the index writer.")
 
+        class FieldIndex:
```
Comment (rstuart85) [inline caterpillar/processing/index.py]: We should use __slots__ here.

File: caterpillar/processing/index.py [hunk 4]
```diff
@@ -338,33 +344,37 @@ def flush(self):
...
+            def __init__(self):
+                self.positions = {}  # Inverted term positions index:: term -> [(start, end,), (star,end,), ...]
+                self.associations = {}  # Inverted term co-occurrence index:: term -> other_term -> count
+                self.frequencies = nltk.probability.FreqDist()  # Inverted term frequency index:: term -> count
```
Comment (rstuart85) [inline caterpillar/processing/index.py]: FreqDist is really really slow. Is there something better we can use here while we are doing this?

Comment (rstuart85) [inline caterpillar/processing/index.py]: Maybe `collections.Counter`?

Comment (krisrogers) [inline caterpillar/processing/index.py]: Yes, although a defaultdict would probably be faster. And probably fastest, bug ugliest:
`try: frequencies[term] += 1 ... except KeyError: frequencies[term] = 1`

Comment (rstuart85) [inline caterpillar/processing/index.py]: I'm fine with that.

**Comment by rstuart85** [comment]
The old approach to metadata doesn't work with this new index per field code. We need to create an index per field regardless of whether it is a text field or not. Implemented this and merging this changeset will also mean updating the topic code. Lets park this for now until we get a chance to make these changes. It might be a job for Sam.

**Pull request closed without merging**